docs/dev_guide.md

```diff
@@ -1,31 +1,166 @@
 # Dev Guide
 
-## Adding a new param in GCSFuse
+## Adding new features to GCSFuse
+
+**If you're interested in adding a new feature, please open a GitHub issue to
+discuss your proposal with the GCSFuse team first before making the
+contribution.**
+
+Rollout of any new feature should be done in 2 phases:
+
+**1. Experimental Phase:**
+
+* New features should initially be introduced behind an experimental flag. These
+  flags should be prefixed with `--experimental-` and marked as hidden in the
+  help output.
+* This signals to users that the feature is in development and may change or be
+  removed.
+* **Testing Requirements**:
+    * Thorough testing is a crucial requirement for any new features added to
+      GCSFuse even during the experimental phase. This helps ensure the
+      feature's stability, identify potential issues early, and prevent
+      regressions.
+    * Add unit tests for all new code written for the experimental feature. This
+      includes testing edge cases, boundary conditions, and error handling.
+    * Add composite tests to validate GCSFuse's end user functionality while
+      isolating the testing to the GCSFuse code itself. This allows for faster
+      and more controlled testing without relying on a real Google Cloud Storage
+      bucket.
+    * Add end-to-end tests for complete feature flows, simulating real-world use
+      with an actual Google Cloud Storage bucket.
+    * For more details on testing best practices, please refer to
+      the [testing guide](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/dev_guide.md#testing-of-new-features)
+      below.
+
+* After the feature development is complete, developers should open a GitHub
+  issue (
+  see [example](https://github.com/GoogleCloudPlatform/gcsfuse/issues/793)) to
+  track the feature's progress towards production readiness. This issue should
+  outline the motivation for the feature, its design, and any known limitations.
+
+**2. Production Rollout:**
+
+* Once the feature has been thoroughly vetted in test environments and approved
+  by the GCSFuse team, it will be rolled out for general availability. The
```

Comment:
Once the feature has been thoroughly vetted in test environments and approved
  by the GCSFuse team, it will be rolled out for general availability. There is no guarantee that experimental features will become generally available.